plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/expressions.kt

```diff
@@ -201,10 +213,79 @@ fun JKVariable.hasUsages(scope: JKTreeElement, context: NewJ2kConverterContext):
 
 context(KtAnalysisSession)
 fun JKVariable.hasWritableUsages(scope: JKTreeElement, context: NewJ2kConverterContext): Boolean =
-    findUsages(scope, context).any {
+    countWritableUsages(scope, context).isNotEmpty()
+
+context(KtAnalysisSession)
+private fun JKVariable.countWritableUsages(scope: JKTreeElement, context: NewJ2kConverterContext): List<JKElement> =
+    findUsages(scope, context).filter {
         it.asAssignmentFromTarget() != null
-                || it.isInDecrementOrIncrement()
+                || it.isInDecrementOrIncrement() || it.asQualifiedAssignmentFromTarget() != null || it.asParenthesizedAssignmentFromTarget() != null
+    }.distinct()
+
+private fun assignmentInLoop(elements: List<JKElement>): Boolean =
+    elements.any { element ->
+        element.parentOfType<JKForInStatement>() != null || element.parentOfType<JKWhileStatement>() != null
+    }
+
+private fun assignmentsOnlyInIfElseConditionals(elements: List<JKElement>): Boolean {
+    val ifElseStatements = elements.map { it.parentOfType<JKIfElseStatement>() }.distinct()
+    if (ifElseStatements.any { it == null }) {
+        return false
+    }
+    if (ifElseStatements.size == 1) {
+        return true
+    }
+    // either separate if/else statements OR if -> else if statement
+    val parentStatement = ifElseStatements.map {
+        var parent = it
+        while (parent?.parentOfType<JKIfElseStatement>() != null) {
+            parent = parent.parentOfType<JKIfElseStatement>()
+        }
+        parent
+    }.distinct()
+    if (parentStatement.size != 1) {
+        return false
     }
+    // ensure no nested statements
+    return ifElseStatements.all { parentStatement[0]?.psi?.text == it?.psi?.text || parentStatement[0]?.psi?.text?.contains("else " + it?.psi?.text) == true }
```

Comment:
Again, access to psi just for the text looks suspicious. Is it really necessary?